projects/baby_driver/test/test_i2c_write.c

```diff
@@ -0,0 +1,387 @@
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_status);
+
+  // Test invalid port for command message
+  command_port = 2;
+  command[1] = command_port;
+
+  // Send CAN message with command message information
+  CAN_TRANSMIT_BABYDRIVER(command[0], command[1], command[2], command[3], command[4], command[5],
+                          command[6], command[7]);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(3, s_times_callback_called);
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, s_received_status);
+
+  // Setup i2c_write_init() with shorter timeout to speed up tests
+  s_timeout_test = true;
+}
+
+// Test whether timeout error is called when no message is returned
+void test_timeout_error_i2c(void) {
+  I2CPort command_port = I2C_PORT_1;
+  I2CAddress command_address = 0x74;
+  uint8_t command_tx_len = 8;
+  uint8_t command_is_reg = 2;
+  uint8_t command_reg = 3;
+  uint8_t command[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND,
+                         command_port,
+                         command_address,
+                         command_tx_len,
+                         command_is_reg,
+                         command_reg,
+                         0,
+                         0 };
+
+  // Test delay between 1st and 2nd data message
+  uint8_t data[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_DATA, 10, 22, 5, 4, 3, 76, 120 };
+
+  // Send CAN message with command message information
+  CAN_TRANSMIT_BABYDRIVER(command[0], command[1], command[2], command[3], command[4], command[5],
+                          command[6], command[7]);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  // Send CAN message with data to be written over i2c
+  CAN_TRANSMIT_BABYDRIVER(data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7]);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  // Delay before next message to trigger timeout error
+  delay_ms(TEST_TIMEOUT_PERIOD_MS);
+  // MS_TEST_HELPER_CAN_TX_RX will stall waiting for a message to be received. This makes it
+  // difficult to check whether the timeout is being triggered on time. By calculating the time it
+  // takes for the status message to be returned, this can be prevented.
+  clock_t start_time = clock();
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  clock_t end_time = clock();
+  // Stores time taken to receive a status message in milliseconds and truncates the number to 0
+  // decimals
+  uint32_t time_taken = 1000 * ((end_time - start_time) / (double)CLOCKS_PER_SEC);
+
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);
+  TEST_ASSERT_EQUAL(STATUS_CODE_TIMEOUT, s_received_status);
+  // It is expected that it will take 10ms +/- 10ms for status to be returned
+  TEST_ASSERT_INT32_WITHIN(10, 10, time_taken);
```

Comment:
This timing-checking stuff seems like it's liable to fail inconsistently, I think it's fine to just check that the timeout was triggered after the timeout period + a bit.